src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -1239,6 +1239,57 @@ tx_gateway_frontend::get_all_transactions() {
       });
 }
 
+ss::future<tx_errc> tx_gateway_frontend::delete_partition_from_tx(
+  kafka::transactional_id tid, tm_transaction::tx_partition ntp) {
+    auto shard = _shard_table.local().shard_for(model::tx_manager_ntp);
+
+    if (shard == std::nullopt) {
+        vlog(txlog.warn, "can't find a shard for {}", model::tx_manager_ntp);
+        co_return tx_errc::shard_not_found;
+    }
+
+    co_return co_await container().invoke_on(
+      *shard,
+      _ssg,
+      [tid, ntp](tx_gateway_frontend& self) -> ss::future<tx_errc> {
+          auto partition = self._partition_manager.local().get(
+            model::tx_manager_ntp);
+          if (!partition) {
+              vlog(
+                txlog.warn,
+                "can't get partition by {} ntp",
+                model::tx_manager_ntp);
+              co_return tx_errc::invalid_txn_state;
+          }
+
+          auto stm = partition->tm_stm();
+
+          if (!stm) {
+              vlog(
+                txlog.warn,
+                "can't get tm stm of the {}' partition",
+                model::tx_manager_ntp);
+              co_return tx_errc::invalid_txn_state;
+          }
+
+          auto gate_lock = gate_guard(self._gate);
+          auto read_lock = co_await stm->read_lock();
```

Comment:
Why are you using `read_lock`, but not `get_tx_lock`?

Reply:
Nice catch. I should you `with()` logic